File: client/news-db/src/Stats.js
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import './Stats.css'; // Ensure you have a CSS file for styling

const Stats = () => {
    const [statsData, setStatsData] = useState([]); // State to store the stats
    const [isLoading, setIsLoading] = useState(true); // Loading state
    const [error, setError] = useState(""); // Error state for API errors

    // Fetch stats from API when the component mounts
    useEffect(() => {
        const fetchStats = async () => {
            try {
                const response = await axios.get('http://localhost:8003/stats');
                setStatsData(response.data); // Assuming the response is an array of stats
            } catch (err) {
                setError('Failed to load stats');
                console.error(err);
            } finally {
                setIsLoading(false);
            }
        };

        fetchStats();
    }, []);

    // Function to get top 10 and remaining items
    const getTop10AndRemaining = (histogramData) => {
        if (!Array.isArray(histogramData)) return { top10: [], remainingCount: 0 };

        const sortedData = histogramData
            .sort((a, b) => b.count - a.count) // Sort by count in descending order
            .slice(0, 10); // Get the top 10

        const remainingCount = histogramData.length - sortedData.length;
        return { top10: sortedData, remainingCount };
    };

    return (
        <div className="stats-container">
            <h2>Statistics</h2>

            {isLoading ? (
                <p>Loading stats...</p>
            ) : error ? (
                <p>{error}</p>
            ) : (
                <div className="stats-cards">
                    {statsData.map((stat, index) => (
                        <div key={index} className="stat-card">
                            <h3>{stat.title}</h3>
                            <p><strong>Words:</strong> {stat.words}</p>
                            <p><strong>Groups:</strong> {stat.groups}</p>
                            <p><strong>Lines:</strong> {stat.lines}</p>
                            <p><strong>Paragraphs:</strong> {stat.paragraphs}</p>

                            {/* Groups Histogram */}
                            <div className="histogram">
                                <h4>Groups Histogram:</h4>
                                <ul>
                                    {getTop10AndRemaining(stat.groups_histogram).top10.map((item, idx) => (
                                        <li key={idx}>
                                            {item.term}: {item.cnt}
                                        </li>
                                    ))}
                                </ul>
                                {getTop10AndRemaining(stat.groups_histogram).remainingCount > 0 && (
                                    <p>And {getTop10AndRemaining(stat.groups_histogram).remainingCount} more...</p>
                                )}

                                {/* Words Histogram */}
                                <h4>Words Histogram:</h4>
                                <ul>
                                    {getTop10AndRemaining(stat.words_histogram).top10.map((item, idx) => (
                                        <li key={idx}>
                                            {item.term}: {item.cnt}
                                        </li>
                                    ))}
                                </ul>
                                {getTop10AndRemaining(stat.words_histogram).remainingCount > 0 && (
                                    <p>And {getTop10AndRemaining(stat.words_histogram).remainingCount} more...</p>
                                )}
                            </div>
                        </div>
                    ))}
                </div>
            )}
        </div>
    );
};

export default Stats;

File: ui/news_db/lib/stats.dart
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

class Stats extends StatefulWidget {
  @override
  _StatsState createState() => _StatsState();
}

class _StatsState extends State<Stats> {
  List<dynamic> statsData = [];
  bool isLoading = true;
  String error = '';

  Future<void> fetchStats() async {
    try {
      final response = await http.get(Uri.parse('http://localhost:8003/stats'));
      if (response.statusCode == 200) {
        setState(() {
          statsData = json.decode(response.body);
        });
      } else {
        setState(() {
          error = 'Failed to load stats';
        });
      }
    } catch (e) {
      setState(() {
        error = 'Failed to load stats';
      });
      print(e);
    } finally {
      setState(() {
        isLoading = false;
      });
    }
  }

  @override
  void initState() {
    super.initState();
    fetchStats();
  }

  Map<String, dynamic> getTop10AndRemaining(List<dynamic> histogramData) {
    if (histogramData == null || histogramData.isEmpty) {
      return {'top10': [], 'remainingCount': 0};
    }

    histogramData.sort((a, b) => b['cnt'].compareTo(a['cnt']));
    final top10 = histogramData.take(10).toList();
    final remainingCount = histogramData.length - top10.length;

    return {'top10': top10, 'remainingCount': remainingCount};
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Statistics')),
      body: Padding(
        padding: const EdgeInsets.all(16.0),
        child: isLoading
            ? Center(child: CircularProgressIndicator())
            : error.isNotEmpty
            ? Center(child: Text(error, style: TextStyle(color: Colors.red)))
            : ListView.builder(
          itemCount: statsData.length,
          itemBuilder: (context, index) {
            final stat = statsData[index];
            final groupsHistogram = getTop10AndRemaining(stat['groups_histogram']);
            final wordsHistogram = getTop10AndRemaining(stat['words_histogram']);

            return Card(
              margin: EdgeInsets.symmetric(vertical: 8),
              elevation: 4,
              child: Padding(
                padding: const EdgeInsets.all(16.0),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(stat['title'], style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold)),
                    SizedBox(height: 10),
                    Text('Words: ${stat['words']}'),
                    Text('Groups: ${stat['groups']}'),
                    Text('Lines: ${stat['lines']}'),
                    Text('Paragraphs: ${stat['paragraphs']}'),
                    SizedBox(height: 15),
                    Text('Groups Histogram:', style: TextStyle(fontWeight: FontWeight.bold)),
                    ...groupsHistogram['top10'].map((item) => Text('${item['term']}: ${item['cnt']}')).toList(),
                    if (groupsHistogram['remainingCount'] > 0)
                      Text('And ${groupsHistogram['remainingCount']} more...'),
                    SizedBox(height: 10),
                    Text('Words Histogram:', style: TextStyle(fontWeight: FontWeight.bold)),
                    ...wordsHistogram['top10'].map((item) => Text('${item['term']}: ${item['cnt']}')).toList(),
                    if (wordsHistogram['remainingCount'] > 0)
                      Text('And ${wordsHistogram['remainingCount']} more...'),
                  ],
                ),
              ),
            );
          },
        ),
      ),
    );
  }
}
